### PR TITLE
Explain Kubevirt

### PR DESCRIPTION
**Pull request by xpivarc**
**What this PR does / why we need it**:
Enable validation on all kubevirt objects.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
KubeVirt is now explainable!
```

### CODE REVIEW

File: tests/kubectl_test.go [hunk 1]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
 	},
 		table.Entry("[rfe_id:3810]explain vm", "vm"),
```
Comment (ILpinto) [inline tests/kubectl_test.go:41]: test_id

File: tests/kubectl_test.go [hunk 2]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
...
 		table.Entry("[rfe_id:3811]explain vmi", "vmi"),
```
Comment (ILpinto) [inline tests/kubectl_test.go:42]: test_id

File: tests/kubectl_test.go [hunk 3]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
...
-		table.PEntry("explain vmim", "vmim"),
-		table.PEntry("explain kv", "kv"),
-		table.PEntry("explain vmsnapshot", "vmsnapshot"),
-		table.PEntry("explain vmsnapshotcontent", "vmsnapshotcontent"),
+		table.Entry("explain vmim", "vmim"),
```
Comment (ILpinto) [inline tests/kubectl_test.go:43]: Add test_id:5178

File: tests/kubectl_test.go [hunk 4]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
...
+		table.Entry("explain kv", "kv"),
```
Comment (ILpinto) [inline tests/kubectl_test.go:44]: Add test_id:5179

File: tests/kubectl_test.go [hunk 5]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
...
+		table.Entry("explain vmsnapshot", "vmsnapshot"),
```
Comment (ILpinto) [inline tests/kubectl_test.go:45]: Add test_id:5180

File: tests/kubectl_test.go [hunk 6]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
...
+		table.Entry("explain vmsnapshotcontent", "vmsnapshotcontent"),
```
Comment (ILpinto) [inline tests/kubectl_test.go:46]: add test_id:5181

File: tests/kubectl_test.go [hunk 7]
```diff
@@ -40,13 +40,13 @@ var _ = Describe("oc/kubectl integration", func() {
...
 	)
 
-	PIt("vmipreset have validation", func() {
+	It("vmipreset have validation", func() {
```
Comment (ILpinto) [inline tests/kubectl_test.go:49]: Add test_id:5182

File: tests/kubectl_test.go [hunk 8]
```diff
@@ -55,7 +55,7 @@ var _ = Describe("oc/kubectl integration", func() {
 		Expect(output).To(ContainSubstring("spec	<Object>"))
 	})
 
-	PIt("vmirs have validation", func() {
+	It("vmirs have validation", func() {
```
Comment (ILpinto) [inline tests/kubectl_test.go:58]: Add test_id:5183

File: tests/operator_test.go [hunk 9]
```diff
@@ -375,6 +375,17 @@ var _ = Describe("[Serial]Operator", func() {
...
+				kv, err := virtClient.KubeVirt(flags.KubeVirtInstallNamespace).Get(kv.Name, &metav1.GetOptions{})
+				ExpectWithOffset(1, err).NotTo(HaveOccurred())
+				kv.Spec.Infra = infra
+				EventuallyWithOffset(1, func() error {
+					_, err := virtClient.KubeVirt(flags.KubeVirtInstallNamespace).Update(kv)
+					return err
+				}, 10*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+				return
+			}
+
```
Comment (davidvossel) [inline tests/operator_test.go:388]: why is this necessary? same with the patchKvWorkloads change. Maybe we should be creating an "Add" patch instead of a replace?

File: hack/builder/Dockerfile [hunk 10]
```diff
@@ -62,9 +62,12 @@ RUN set -x && \
     go get -v github.com/golang/protobuf/protoc-gen-go@1643683 && \
     go get -v k8s.io/code-generator/cmd/client-gen@v0.16.4 && \
     go get -v github.com/securego/gosec/v2/cmd/gosec@0ce48a5 && \
-    go get -v sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.0 && \
     go clean -cache -modcache
 
+RUN git clone https://github.com/xpivarc/controller-tools.git && \
+    cd controller-tools/cmd/controller-gen && git checkout string-fix && \
+    source /etc/profile.d/gimme.sh && go install 
+
```
Comment (davidvossel) [inline hack/builder/Dockerfile:70]: i see you made a bug fix that you're trying to get merged into controller-tools https://github.com/kubernetes-sigs/controller-tools/issues/507

What field or fields does this bug fix impact in our API?

File: tests/operator_test.go (hunk 9 above)
Comment (xpivarc) [reply tests/operator_test.go:388]: The test is trying to remove the infra/workload field. With patch, it actually set it to null but that field is not nullable. I guess we could use remove?

File: hack/builder/Dockerfile (hunk 10 above)
Comment (xpivarc) [reply hack/builder/Dockerfile:70]: Fields that have 'string' in json tag. [i.e](https://github.com/kubevirt/kubevirt/pull/4396/files?file-filters%5B%5D=.go&file-filters%5B%5D=.sh&file-filters%5B%5D=No+extension#diff-194ad1a8d11fa7160d03e60a5ab73fef10057b639474874d7f5ee32040b40cb3R270).
`integer` -> `string` so validation will not fail.

File: tests/operator_test.go (hunk 9 above)
Comment (davidvossel) [reply tests/operator_test.go]: oh, i miss read the intent. Yeah, try remove. If that doesn't work, you could replace the entire spec field if you needed.

File: hack/builder/Dockerfile (hunk 10 above)
Comment (davidvossel) [reply hack/builder/Dockerfile:70]: i see. so would this make both  `resource.requests.CPU = 2` and `resource.requests.CPU = "2"` work? where the latter would fail without the change?

Comment (rmohr) [reply hack/builder/Dockerfile:70]: What I don't understand, if we want both to be valid, why don't we use `IntOrString` type?

Comment (xpivarc) [reply hack/builder/Dockerfile:70]: Only the latter succeeds with this change. If we want both to work we need `IntOrString` as Roman suggested. Maybe I misunderstood the API, do you think we want both?

**Review by rmohr — CHANGES_REQUESTED** [review]
/hold

The openapi spec must not change in backward incompatible ways. Booleans need to remain boolean.

**Comment by xpivarc** [comment]
I will need to rethink [this](https://github.com/kubevirt/kubevirt/blob/09c2b1a9e71b24bc22357fa64639f947bbb91f59/staging/src/kubevirt.io/client-go/apis/snapshot/v1alpha1/types.go#L170) specific field which is blocking this PR. A possible solution would be to change it to `PersistentVolumeClaimVolumeSource` but I am not sure of all implications right now.

**Comment by davidvossel** [comment]
> Update: We need to keep most of the PersistentVolumeClaims fields, will create a 'mirror' struct to overcome this issue.

@mhenriks ^ just making sure you're aware and part of this discussion.

**Review by Donkey23 — APPROVED** [review]

**Comment by davidvossel** [comment]
@xpivarc anything left to do here?

**Comment by xpivarc** [comment]
> @xpivarc anything left to do here?

Cluster-up was broken. It should be fixed and ready to go.

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 11]
```diff
@@ -1503,9 +1503,7 @@ type DeveloperConfiguration struct {
 // PermittedHostDevices holds inforamtion about devices allowed for passthrough
 // +k8s:openapi-gen=true
 type PermittedHostDevices struct {
-	// +listType=set
-	PciHostDevices []PciHostDevice `json:"pciHostDevices,omitempty"`
-	// +listType=set
+	PciHostDevices  []PciHostDevice      `json:"pciHostDevices,omitempty"`
 	MediatedDevices []MediatedHostDevice `json:"mediatedDevices,omitempty"`
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1507]: if `set` is the wrong listType here, what should we be marking it as? `listType=atomic` ?

**Review by davidvossel — COMMENTED** [review]
/approve

I just had one comment about the change to remove listType=set that I'd like to get clarity on before we remove the hold

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 11 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1507]: Yes or `listType=map`. `atomic` would always replace whole slice and with `map` you provide key based on which the update will be performed.

Comment (ashleyschuett) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1507]: should this be updated? instead of adding it as an "API rule violation:" to the api-rules-validations-* lists?

Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1507]: It is not necessary for this PR and unless we decide to support SSA. This can be 'fixed' anytime in future.

**Comment by xpivarc** [comment]
@rmohr  issues were addressed, pinging if you want to have second look

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 11 above)
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1507]: If we're going to change it in this PR, we can update it in this PR instead of adding a new violation. listType=atomic should be fine.
